### ISSUE TITLE
WarningsNotAsErrors Intellisense Doesn't Match Build

### ISSUE THREAD

**Issue by owenneil** [dotnet/project-system#3680]
When using the "WarningsNotAsErrors" feature Intellisense errors will frequently get out of sync with the actual build, showing with red squiggles and in the errors list when build is successful. I've not been able to figure out the exact combination of conditions, and often simply adding or removing some of the warnings from the list will stop the behavior. I've attached a simple project that builds with warnings but shows the issue is VS. 
[WarningTest.zip](https://github.com/dotnet/project-system/files/2147179/WarningTest.zip)

**Comment by Pilchie** [comment]
@davkean @jasonmalinowski @heejaechang can someone take a look and see where things are going wrong here?

**Comment by heejaechang** [comment]
are we talking about warning as error flag? what is warning as not error? anyway, so are we propagating this from project system to roslyn after the initial setting?

**Comment by owenneil** [comment]
To clarify, the "WarningsNotAsErrors" property allows you to set warnings as errors, but have a specific set still be warnings. It's the opposite of putting a list of warnings in WarningsAsErrors (a whitelist vs blacklist). I've got a project that I'm trying to enable most warnings as errors along with Stylecop and Fxcop analyzers, but trying to fix all the warnings in one go would be difficult. So I turned on warnings as errors, then disabled a bunch using this setting. I could use NoWarn, but then they aren't visible at all, so it is difficult to see fix them incrementally as I touch code and even more instances tend to be introduced unknowingly. It sort of works, but once you add several warnings it just stops and requires lots of fiddling to get it to work again for intellisense.

**Comment by Pilchie** [comment]
@owenneil what type of project are you working with?  Is it an sdk-style project (.NET Core/.NET Standard), or an original one?

**Comment by owenneil** [comment]
It's a SDK style project, the sample attached is building .Net Core, but my real project is framework (with a SDK style project). 

The project file attached is really small, just this:
```
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
    <TargetFramework>netcoreapp2.1</TargetFramework>
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
    <WarningsNotAsErrors>CS0169,CS0169,SA1633,SA1200,SA1600,IDE1006,SA1307,CS1591,SA1306,SA1202,SA1310,SA1401,SA1025,SA1522,SA1502,SA1500,SA1210</WarningsNotAsErrors>
    <GenerateDocumentationFile>true</GenerateDocumentationFile>
   </PropertyGroup>
  <ItemGroup>
    <PackageReference Include="StyleCop.Analyzers" Version="1.0.2" />
  </ItemGroup>
</Project>
```

**Comment by Pilchie** [comment]
@owenneil Can you run VS with `PROJECTSYSTEM_PROJECTOUTPUTPANEENABLED=1`, and there should be an extra output window pane with logging that will tell us what the project system passes to the language service?

**Comment by owenneil** [comment]
@Pilchie did that, I'll attach the full output, but here are the command line args, it does seem to pass those flags correctly. 
`/checked- /out:obj\Debug\netcoreapp2.1\WarningTest.dll /warnaserror-:CS0169,CS0169,SA1633,SA1200,SA1600,IDE1006,SA1307,CS1591,SA1306,SA1202,SA1310,SA1401,SA1025,SA1522,SA1502,SA1500,SA1210 obj\Debug\netcoreapp2.1\\TemporaryGeneratedFile_E7A71F73-0F8D-4B9B-B56E-8E70B10BC5D3.cs /nowarn:1701,1702,1701,1702,2008 /filealign:512 /doc:obj\Debug\netcoreapp2.1\WarningTest.xml obj\Debug\netcoreapp2.1\\TemporaryGeneratedFile_5937a670-0e60-4077-877b-f7221da3dda1.cs /noconfig /warnaserror+ /debug+ /warnaserror+:NU1605 /define:TRACE;DEBUG;NETCOREAPP;NETCOREAPP2_1 obj\Debug\netcoreapp2.1\\TemporaryGeneratedFile_036C0B5B-1481-4323-8D20-8F5ADCB23D92.cs /debug:portable C:\Users\neowen\AppData\Local\Temp\.NETCoreApp,Version=v2.1.AssemblyAttributes.cs Class1.cs /errorreport:prompt /optimize- /nostdlib+ /deterministic+ /utf8output /target:library /warn:4 /unsafe- /errorendlocation /ruleset:"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Team Tools\Static Analysis Tools\\Rule Sets\MinimumRecommendedRules.ruleset" obj\Debug\netcoreapp2.1\WarningTest.AssemblyInfo.cs /preferreduilang:en-US
`
Here is the full command line arguments from the build output, to compare:
`/noconfig /unsafe- /checked- /nowarn:1701,1702,1701,1702,2008 /nostdlib+ /errorreport:prompt /warn:4 /doc:obj\Debug\netcoreapp2.1\WarningTest.xml /define:TRACE;DEBUG;NETCOREAPP;NETCOREAPP2_1 /errorendlocation /preferreduilang:en-US /reference:"C:\Program Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.app\2.1.0\ref\netcoreapp2.1\Microsoft.CSharp.dll"` ---snip out many references to these---`Files\dotnet\sdk\NuGetFallbackFolder\microsoft.netcore.app\2.1.0\ref\netcoreapp2.1\WindowsBase.dll" /debug+ /debug:portable /filealign:512 /optimize- /out:obj\Debug\netcoreapp2.1\WarningTest.dll /ruleset:"C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\Team Tools\Static Analysis Tools\\Rule Sets\MinimumRecommendedRules.ruleset" /target:library /warnaserror+ /utf8output /deterministic+ /analyzer:C:\Users\neowen\.nuget\packages\stylecop.analyzers\1.0.2\analyzers\dotnet\cs\StyleCop.Analyzers.CodeFixes.dll /analyzer:C:\Users\neowen\.nuget\packages\stylecop.analyzers\1.0.2\analyzers\dotnet\cs\StyleCop.Analyzers.dll Class1.cs "C:\Users\neowen\AppData\Local\Temp\.NETCoreApp,Version=v2.1.AssemblyAttributes.cs" obj\Debug\netcoreapp2.1\\TemporaryGeneratedFile_E7A71F73-0F8D-4B9B-B56E-8E70B10BC5D3.cs obj\Debug\netcoreapp2.1\\TemporaryGeneratedFile_036C0B5B-1481-4323-8D20-8F5ADCB23D92.cs obj\Debug\netcoreapp2.1\\TemporaryGeneratedFile_5937a670-0e60-4077-877b-f7221da3dda1.cs obj\Debug\netcoreapp2.1\WarningTest.AssemblyInfo.cs /warnaserror+:NU1605 /warnaserror-:CS0169,CS0169,SA1633,SA1200,SA1600,IDE1006,SA1307,CS1591,SA1306,SA1202,SA1310,SA1401,SA1025,SA1522,SA1502,SA1500,SA1210`


[ProjectOutput.txt](https://github.com/dotnet/project-system/files/2156138/ProjectOutput.txt)
[BuildOutPut.txt](https://github.com/dotnet/project-system/files/2156142/BuildOutPut.txt)

**Comment by owenneil** [comment]
Hmm, I did a bit more fiddling and managed to get the issue to go away in my bigger project by: 
- Removing all duplicate warnings listed.
- Fixing two warnings that were separated by semicolon rather than comma mistakenly. 

Not sure why these issues are treated differently in build vs intellisense, which makes this confusing, but I'm glad I can at least stop it.

**Comment by Pilchie** [comment]
This issue was moved to dotnet/roslyn#28262

**Comment by davkean** [comment]
Repo'ing this as the underling cause is the project system: https://github.com/dotnet/roslyn/issues/28262#issuecomment-403322803

**Comment by RikkiGibson** [comment]
@jjmew @davkean is there any chance this could get picked up? We just got another bug report in Roslyn that is likely another instance of this bug.

**Comment by jjmew** [comment]
@RikkiGibson it is unlikely that we can get to this one soon. If you have some information to help us re-evaluate the priority  then let us know. We will be more than happy to figure it out.

**Comment by TFTomSun** [comment]
I observe a similar problem. When I set:
```xml

    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
    <!--CS0612 is for obsolete warnings that don't have a message-->
    <WarningsNotAsErrors>CS0618;CS1591</WarningsNotAsErrors>

```
in the Directory.Build.props file, the XML comment warnings are shown as errors in the editor and in the error list. I think that haven't been always like that. I just discovered it with the latest VS 2019 preview.
![image](https://user-images.githubusercontent.com/5387192/111030052-80297880-8400-11eb-89f9-317a3dc60510.png)

**Comment by TFTomSun** [comment]
The workaround that I just found is to add the warnings for which I don't want to get errors to the NoWarn property:
```xml
    <TreatWarningsAsErrors>true</TreatWarningsAsErrors>
    <!--CS0612 is for obsolete warnings that don't have a message-->
    <WarningsNotAsErrors>CS0618;CS1591</WarningsNotAsErrors>
    <NoWarn>CS0618;CS1591</NoWarn>
```
Of course, these warnings are then not listed as warnings anymore, too. But better than getting the error list messed up.

**Comment by jcouv** [comment]
@davkean Any chance we could address this?

More customers have hit this. @RikkiGibson and I have been debugging an issue with `<WarningsAsErrors>nullable</WarningsAsErrors>` not taking effect in the IDE (Error List and squiggles) (issue https://github.com/dotnet/roslyn/issues/52407) and it looks like the compiler is given arguments in different order depending on exactly what diagnostics are listed in `WarningsAsErrors`. 
Sometimes a `warnaserror-` gets moved after `warnaserror+:...` (with specific diagnostic IDs) which causes a different behavior from the compiler. 
This is extremely surprising and pretty painful to troubleshoot :-/

**Comment by davkean** [comment]
I don't work on this anymore, I've unassigned myself and sent back to the triage team.

**Comment by jcouv** [comment]
@kendrahavens @mikadumont @jjmew @jinujoseph This is an issue that's affecting IDE scenarios. Could you help prioritize it?

TL;DR: the compiler is sensitive to ordering of command-line arguments (in this case different `warnaserror` values), but the process of making a compiler for the IDE effectively shuffles some of the arguments. This leads to very surprising and unpredictable behavior (diagnostics treated as errors or not depending on irrelevant changes to the project file).

**Comment by jjmew** [comment]
@ocalles Could you take a look at this? This most likely will involve CPS.

**Comment by RikkiGibson** [comment]
IIRC, the root cause is basically that an interface in internal code expects an ordered dictionary, but the implementation uses an unordered dictionary.

**Comment by drewnoakes** [comment]
Yes, CPS's "rule" processing stores items in an `ImmutableDictionary<>` which orders items by their hash codes, essentially randomizing their order.

Options are:

1. use the raw build data via `ProjectBuildSnapshotService` which retains item order, or
2. change CPS to retain order by backing the `IImmutableDictionary<,>` with a type that preserves order.

**Comment by jaredpar** [comment]
The PR in question does not address this issue. Leaving a comment on the PR.

**Comment by jcouv** [comment]
I think this was fixed by https://github.com/dotnet/project-system/pull/7174
I'll let @ocallesp verify and close. Thanks

**Comment by MiYanni** [comment]
@ocalles Can you verify this is fixed? Thanks!

**Comment by ocallesp** [comment]
Does not repro.

It was fixed by #7174
